yarpl/include/yarpl/observable/ObservableOperator.h

```diff
@@ -481,6 +481,73 @@ class SubscribeOnOperator
   std::unique_ptr<Worker> worker_;
 };
 
+template <typename T, typename GeneratorClass>
+class IteratorObservable : public Observable<T> {
+public:
+  Reference<Subscription> subscribe(Reference<Observer<T>> observer) override {
+    static_assert(
+      std::is_base_of<
+        IteratorObservable,
+        GeneratorClass
+      >::value,
+      "Inherit from IteratorObservable like `MyClass : public IteratorObservable<T, MyClass>`"
+    );
+
+    // can be stack allocated; we know nothing escapes
+    GeneratorClass instantiated(*static_cast<GeneratorClass const*>(this));
+    // so we can call protected methods on IteratorObservable even if they were
+    // made private in the subclass
+    IteratorObservable* withvtable = &instantiated;
+
+    auto subscription = make_ref<IteratorObservableSubscription>();
+
+    withvtable->beforeOnSubscribe();
+    observer->onSubscribe(subscription);
+
+    while(!subscription->isCancelled() && withvtable->haveNext()) {
+      observer->onNext(withvtable->getNext());
+    }
+    observer->onComplete();
+    withvtable->afterOnComplete();
+
+    return subscription;
+  }
+
+private:
+  struct IteratorObservableSubscription : public Subscription {
```

Comment:
what is this subclass for? Subscription::cancel() does exactly what your implementation of cancel does...